### PR TITLE
Add simulated mode support

### PR DESCRIPTION
**Pull request by mdheri**
Adding support for simulated mode and compatibility feature flags

Asana: https://app.asana.com/0/1202541719267260/1202555728616891/f

### CODE REVIEW

File: lib/smartcar.rb [hunk 1]
```diff
@@ -163,23 +166,35 @@ def verify_payload(amt, signature, body)
...
       query_params = {
         vin: vin,
         scope: scope.join(' '),
         country: country
       }
       query_params[:flags] = options[:flags].map { |key, value| "#{key}:#{value}" }.join(' ') if options[:flags]
-      query_params[:mode] = options[:test_mode].is_a?(TrueClass) ? 'test' : 'live' unless options[:test_mode].nil?
+
+      if options[:test_mode]
+        warn '[DEPRECATION] The "testMode" parameter is deprecated, please use the "mode" parameter instead.'
```
Comment (gurpreetatwal) [inline lib/smartcar.rb:179]:
```suggestion
        warn '[DEPRECATION] The "test_mode" parameter is deprecated, please use the "mode" parameter instead.'
```

File: lib/smartcar/auth_client.rb [hunk 2]
```diff
@@ -14,14 +14,26 @@ class AuthClient
...
+    # Should be one of test, live or simulated.
     # @return [Smartcar::AuthClient] Returns a Smartcar::AuthClient Object that has other methods
     def initialize(options)
       options[:redirect_uri] ||= get_config('SMARTCAR_REDIRECT_URI')
       options[:client_id] ||= get_config('SMARTCAR_CLIENT_ID')
       options[:client_secret] ||= get_config('SMARTCAR_CLIENT_SECRET')
-      options[:mode] = options[:test_mode].is_a?(TrueClass) ? TEST : LIVE
+
+      if options[:test_mode]
+        warn '[DEPRECATION] The "testMode" parameter is deprecated, please use the "mode" parameter instead.'
```
Comment (gurpreetatwal) [inline lib/smartcar/auth_client.rb:28]:
```suggestion
        warn '[DEPRECATION] The "test_mode" parameter is deprecated, please use the "mode" parameter instead.'
```

File: lib/smartcar/vehicle.rb [hunk 3]
```diff
@@ -181,7 +183,8 @@ def initialize(token:, id:, options: {})
                         when :delete
                           delete(item[:path].call(id))
                         else
-                          fetch(path: item[:path].call(id))
+                          query_params = @flags.nil? ? {} : { flags: @flags }
```
Comment (gurpreetatwal) [inline lib/smartcar/vehicle.rb:186]: Should add flags to all request types

File: lib/smartcar/vehicle.rb [hunk 4]
```diff
@@ -181,7 +183,8 @@ def initialize(token:, id:, options: {})
...
+                          fetch(path: item[:path].call(id), query_params: query_params)
                         end
         build_aliases(build_response(body, headers), item[:aliases])
       end
```
Comment (gurpreetatwal) [inline lib/smartcar/vehicle.rb:190]: Also need to add the flags into the methods listed below:
- permissions
- subscribe
- unsubscribe
- batch
- request

There might be some fancy way that you can set the flags in one place and have them be picked up by all of the methods. @s-ashwinkumar will know if there is

**Review by gurpreetatwal — COMMENTED** [review]
Overall looks good, added a few comments but will defer the rest of the reviews and approval of the PR to @s-ashwinkumar z

File: lib/smartcar/vehicle.rb (hunk 4 above)
Comment (s-ashwinkumar) [reply lib/smartcar/vehicle.rb:190]: With a little bit of modification to the [define_method block](https://github.com/smartcar/ruby-sdk/blob/master/lib/smartcar/vehicle.rb#L172-L188) we can add this into the[ config that  is used to generate those methods](https://github.com/smartcar/ruby-sdk/blob/master/lib/smartcar/vehicle.rb#L19-L71)

File: lib/smartcar/auth_client.rb [hunk 5]
```diff
@@ -14,14 +14,26 @@ class AuthClient
...
+        options[:mode] = options[:test_mode].is_a?(TrueClass) ? 'test' : 'live'
```
[the remark this answers is not in the document]
Comment (s-ashwinkumar) [reply lib/smartcar/auth_client.rb:29]: I am also ok if you want to do it. Basically it looks like `LIVE` or `Smartcar::LIVE` does not exist we just have to replace its mention with the string live and test instead.

**Review by s-ashwinkumar — CHANGES_REQUESTED** [review]

File: lib/smartcar/utils.rb [hunk 6]
```diff
@@ -127,5 +127,16 @@ def convert_path_to_attribute(path)
 
       path.split('/').reject(&:empty?).join('_').to_sym
     end
+
+    # adds to query_params if value is not nil
+    def add_query_params(query_params, key, value)
```
Comment (s-ashwinkumar) [inline lib/smartcar/utils.rb:132]: remove this function as per discussion!

File: lib/smartcar/utils.rb [hunk 7]
```diff
@@ -127,5 +127,11 @@ def convert_path_to_attribute(path)
 
       path.split('/').reject(&:empty?).join('_').to_sym
     end
+
+    # takes query parameters and returns them as a string
+    # EX - {'country': 'DE', 'flags': true} -> "county:DE flags:true"
+    def stringify_params(query_params)
+      query_params&.map { |key, value| "#{key}:#{value}" }&.join(' ')
```
Comment (s-ashwinkumar) [inline lib/smartcar/utils.rb:134]: good one..with the safe navigator operator !! 👍🏽

**Review by s-ashwinkumar — COMMENTED** [review]
Looks good to me ! There are couple of things that we can discuss before merging 
- I think we should also add a test for a delete and post request with the mode param to make sure the params are passed properly. We have a test for sample get req with the parameter
- This one is nit and i don't mind if we leave this one out for the sake of getting this merged, but we could probably just create a function like `handle_mode_parameter` and move the logic into a function so that it can be reused in both the places ...ie compatibility and vehicle object constructor.

**Review by s-ashwinkumar — APPROVED** [review]
Discussed the two additional tests for subscribe and unsubscribe ot make sure parameters are passed in post and delete requests.

Otherwise good to go 👍🏽

**Pull request merged**